snakemake/sourcecache.py

```diff
@@ -251,6 +262,10 @@ def infer_source_file(path_or_uri, basedir: SourceFile = None):
             return path_or_uri
         else:
             path_or_uri = path_or_uri.get_path_or_uri()
+    if isinstance(basedir, LocalGitFile) and isinstance(path_or_uri, str):
+        local_base = os.path.dirname(basedir.path).lstrip("/")
```

Comment:
The SourceFile class has a get_basedir method.

Reply:
It also has a join method. In fact, I am quite sure that no special handling for a local git file should be necessary. Otherwise, we should rather fix it in the class hierarchy

Reply:
refactored